File: apps/halide/unsharp/unsharp.py
from __future__ import annotations

from exo import *
from exo.libs.memories import DRAM_STACK
from exo.platforms.x86 import *

from exo.stdlib.scheduling import *
from exo.stdlib.halide_scheduling_ops import *
from exo.stdlib.inspection import get_enclosing_loop_by_name
from exo.stdlib.stdlib import vectorize


def halide_tile(p, buffer, y, x, yi, xi, yTile, xTile):
    assign = p.find(f"{buffer} = _")
    y_loop = get_enclosing_loop_by_name(p, assign, y)
    x_loop = get_enclosing_loop_by_name(p, assign, x)

    return tile(p, y_loop, x_loop, [y, yi], [x, xi], yTile, xTile, perfect=True)


def halide_split(p, stage, x, xo, xi, split_factor):
    loop = get_enclosing_loop_by_name(p, p.find(f"{stage} = _"), x)
    return split(p, loop, xo, xi, split_factor)


def halide_compute_at(p, producer: str, consumer: str, loop: str):
    x_loop = get_enclosing_loop_by_name(p, p.find(f"{consumer} = _"), loop)
    return compute_at(p, producer, consumer, x_loop)


def halide_parallel(p, loop: str):
    return parallelize_loop(p, p.find_loop(loop))


avx_f32_insts = [
    mm256_loadu_si256,
    mm256_storeu_si256,
    mm256_add_epi16,
]


def halide_vectorize(p, buffer: str, loop: str, width: int):
    loop = get_enclosing_loop_by_name(p, p.find(f"{buffer} = _"), loop)
    p = vectorize(p, loop, width, "f32", AVX2, avx_f32_insts, rules=[], tail="perfect")

    return p


# TODO: define a special case of compute_at which fully inlines computations
# TODO: implement halide's reorder over arbitrary loop nests


# from math import exp, pi, sqrt
# sigma = 1.5
# kernel = [exp(-x * x / (2 * sigma * sigma)) / (sqrt(2 * pi) * sigma) for x in range(4)]
# print(kernel)
# >> [0.2659615202676218, 0.2129653370149015, 0.10934004978399575, 0.035993977675458706]
# TODO: it would be nice if meta-programming allowed pre-computing values outside of the proc


@proc
def unsharp(W: size, H: size, output: f32[H, W], input: f32[H + 6, W + 6, 3]):
    assert H % 32 == 0
    assert W % 256 == 0

    gray: f32[H + 6, W + 6]
    for y in seq(0, H + 6):
        for x in seq(0, W + 6):
            gray[y, x] = (
                0.299 * input[y, x, 0] + 0.587 * input[y, x, 1] + 0.114 * input[y, x, 2]
            )

    blur_y: f32[H, W + 6]
    for y in seq(0, H):
        for x in seq(0, W + 6):
            blur_y[y, x] = (
                0.2659615202676218 * gray[y + 3, x]
                + 0.2129653370149015 * (gray[y + 2, x] + gray[y + 4, x])
                + 0.10934004978399575 * (gray[y + 1, x] + gray[y + 5, x])
                + 0.035993977675458706 * (gray[y + 0, x] + gray[y + 6, x])
            )

    blur_x: f32[H, W]
    for y in seq(0, H):
        for x in seq(0, W):
            blur_x[y, x] = (
                0.2659615202676218 * blur_y[y, x + 3]
                + 0.2129653370149015 * (blur_y[y, x + 2] + blur_y[y, x + 4])
                + 0.10934004978399575 * (blur_y[y, x + 1] + blur_y[y, x + 5])
                + 0.035993977675458706 * (blur_y[y, x + 0] + blur_y[y, x + 6])
            )

    sharpen: f32[H, W]
    for y in seq(0, H):
        for x in seq(0, W):
            sharpen[y, x] = 2 * gray[y + 3, x + 3] - blur_x[y, x]

    ratio: f32[H, W]
    for y in seq(0, H):
        for x in seq(0, W):
            ratio[y, x] = sharpen[y, x] / gray[y + 3, x + 3]

    for y in seq(0, H):
        for x in seq(0, W):
            for c in seq(0, 3):
                output[y, x] = ratio[y, x] * input[y + 3, x + 3, c]


unsharp = halide_split(unsharp, "output", "y", "y", "yi", 32)
unsharp = halide_compute_at(unsharp, "ratio", "output", "yi")
unsharp = halide_compute_at(unsharp, "sharpen", "ratio", "x")

print(unsharp)

File: apps/halide/unsharp/unsharp/unsharp.c
#include "unsharp.h"

#include <stdio.h>
#include <stdlib.h>

// unsharp(
//     W : size,
//     H : size,
//     output : f32[H, W] @DRAM,
//     input : f32[H + 6, W + 6, 3] @DRAM
// )
void unsharp(void *ctxt, int_fast32_t W, int_fast32_t H, float *output,
    const float *input) {
  EXO_ASSUME(H % 32 == 0);
  EXO_ASSUME(W % 256 == 0);
  float *gray = (float *)malloc((6 + H) * (6 + W) * sizeof(*gray));
  for (int_fast32_t y = 0; y < 6 + H; y++) {
    for (int_fast32_t x = 0; x < 6 + W; x++) {
      gray[y * (6 + W) + x] = 0.299f * input[y * (W + 6) * 3 + x * 3] +
                              0.587f * input[y * (W + 6) * 3 + x * 3 + 1] +
                              0.114f * input[y * (W + 6) * 3 + x * 3 + 2];
    }
  }
  float *blur_y = (float *)malloc(H * (6 + W) * sizeof(*blur_y));
  for (int_fast32_t y = 0; y < H; y++) {
    for (int_fast32_t x = 0; x < 6 + W; x++) {
      blur_y[y * (6 + W) + x] =
          0.2659615202676218f * gray[(3 + y) * (6 + W) + x] +
          0.2129653370149015f *
              (gray[(2 + y) * (6 + W) + x] + gray[(4 + y) * (6 + W) + x]) +
          0.10934004978399575f *
              (gray[(1 + y) * (6 + W) + x] + gray[(5 + y) * (6 + W) + x]) +
          0.035993977675458706f *
              (gray[y * (6 + W) + x] + gray[(6 + y) * (6 + W) + x]);
    }
  }
  float *blur_x = (float *)malloc(H * W * sizeof(*blur_x));
  for (int_fast32_t y = 0; y < H; y++) {
    for (int_fast32_t x = 0; x < W; x++) {
      blur_x[y * W + x] =
          0.2659615202676218f * blur_y[y * (6 + W) + 3 + x] +
          0.2129653370149015f *
              (blur_y[y * (6 + W) + 2 + x] + blur_y[y * (6 + W) + 4 + x]) +
          0.10934004978399575f *
              (blur_y[y * (6 + W) + 1 + x] + blur_y[y * (6 + W) + 5 + x]) +
          0.035993977675458706f *
              (blur_y[y * (6 + W) + x] + blur_y[y * (6 + W) + 6 + x]);
    }
  }
  free(blur_y);
  for (int_fast32_t y = 0; y < ((H) / (32)); y++) {
    for (int_fast32_t yi = 0; yi < 32; yi++) {
      float *ratio = (float *)malloc(1 * W * sizeof(*ratio));
      for (int_fast32_t yii = 0; yii < 1; yii++) {
        for (int_fast32_t x = 0; x < W; x++) {
          float *sharpen = (float *)malloc(1 * 1 * sizeof(*sharpen));
          for (int_fast32_t xi = 0; xi < 1; xi++) {
            for (int_fast32_t yiii = 0; yiii < 1; yiii++) {
              sharpen[yiii + xi] =
                  2.0f * gray[(3 + yi + yii + yiii + 32 * y) * (6 + W) + 3 + x +
                              xi] -
                  blur_x[(yi + yii + yiii + 32 * y) * W + x + xi];
            }
          }
          ratio[yii * W + x] =
              sharpen[0] / gray[(3 + yi + yii + 32 * y) * (6 + W) + 3 + x];
          free(sharpen);
        }
      }
      for (int_fast32_t x = 0; x < W; x++) {
        for (int_fast32_t c = 0; c < 3; c++) {
          output[(yi + 32 * y) * W + x] =
              ratio[x] *
              input[(3 + yi + 32 * y) * (W + 6) * 3 + (3 + x) * 3 + c];
        }
      }
      free(ratio);
    }
  }
  free(blur_x);
  free(gray);
}
